### PR TITLE
feat(plugin-persistence-fabric): add new fabric persistence plugin

### PR DESCRIPTION
**Pull request by barnapa**
Add a new plugin for storing hyperledger fabric ledger data into a database
Add functional tests for plugin and data access layer operations.
operating with fabric-all-in-one docker composition
Tests assume any postgres database, but for final deployment postgres or supabase is assumed.
Data fed by this plugin can later be visualized by a GUI application or analyzed directly. ( right now the GUI will have switch to operate either with Ethereum ledger or Fabric ledger )

Depends on: https://github.com/hyperledger/cacti/pull/2259
Depends on: https://github.com/hyperledger/cacti/pull/2265

Signed-off-by: Barnaba Pawelczak [barnaba.pawelczak@fujitsu.com](mailto:barnaba.pawelczak@fujitsu.com)

### CODE REVIEW

**Review by petermetz — CHANGES_REQUESTED** [review]
@barnapa Please re-organize the commits a little bit to make it easier to review (right now the diff is huge because of the parent PRs in the stack and the 4 commits you have in the PR all look like they belong to your PR itself (but then when I look at the diff I can tell that the parent PRs diff are also included because it has the cmd-gui-app package in too).

If you reorganize your commits so that:
1. There's only one commit with the changes that belong to *this* PR and
2. The other 2 commits come from the 2 parent PRs
then we can review just your specific commit by filtering the diff down to just that (but right now this is not possible).

If you need any help with making this happen (git mechanics or otherwise) then just let me know or join one of the daily pair programming calls that we have and we can screen share: https://wiki.hyperledger.org/display/cactus/Daily+Pair+Programming+Calls

**Review by petermetz — CHANGES_REQUESTED** [review]
> ok it is re organized

@barnapa It is re-organized, but all you did from what I can tell is squash everything together instead of separating the parent PR commits from this PR. You have a single commit that has a 115 files in its diff and it has the GUI app code in it too. Please double check https://github.com/hyperledger/cacti/pull/2331#pullrequestreview-1360124392

**Comment by repository23** [comment]
there are some changes in gui also. 
 (parent GUI covers only ethereum ledger, this gui cover also fabric ledger )
 so make another PR just for changes in GUI ?
 
 changes for plugin ethereum persistence are not included.

**Comment by petermetz** [comment]
> there are some changes in gui also. (parent GUI covers only ethereum ledger, this gui cover also fabric ledger ) so make another PR just for changes in GUI ?
> 
> changes for plugin ethereum persistence are not included.

You could open a separate PR for the GUI changes related to the fabric persistence plugin, yes.
We'll still need to separate out the commits in this PR though so that it is differentiated what goes in with this PR and what goes in with the parent PRs. I don't know how else to explain it other than this plus what I said in https://github.com/hyperledger/cacti/pull/2331#pullrequestreview-1360124392 but I could either show you live or the slower and easier option is to just wait until the parent gets merged and then you won't have to do the commit surgery, you can just do a rebase onto upstream main and all will be well.
I'm fine with either one of those options. You go with your preference depending on how urgent the review on this PR is.

**Comment by barnapa** [comment]
I removed gui for fabric from this pull it will be in another one

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:417]: Can we straightaway get the `tempBlock` var by executing JSON.parse(JSON.stringify(block.data))?

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:483]: Can we straightaway get the `tempBlock` var by executing JSON.parse(JSON.stringify(block.data))?

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:598]: Instead of creating all these vars differently, can we use `InsertDetailedTransactionEntryInterface` here, have all of these referenced by a single var and then use them like var_name.endorser_signature?
And for if else statements we can do something like `if (condition) ? var_name.endorser_signature = ... : var_name.endorser_signature = ""`

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:707]: This should only be executed if we have a valid chaincodeID, because according to this code, chaincodeID only exists when `transactionDataObject.payload.data.actions !== undefined`

And then if `chaincode_id` is valid only for this above condition, then function calls like `insertDetailedTransactionEntry` should also be done only if the above condition is satisfied / exception is handled.

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/web-services/status-endpoint-v1.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-persistence-fabric/src/main/typescript/web-services/status-endpoint-v1.ts:54]: Should we not return the OAS.path here (the commented code)?

File: packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/fabric-setup-helpers.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/fabric-setup-helpers.ts:1]: Why do we have this file under tests? (I don't see it being consumed by any other test files)

File: tsconfig.base.json
Comment (jagpreetsinghsasan) [inline tsconfig.base.json:1]: I can see a lot of really helpful comments being removed from this file. Can these be placed back in, as they are helpful in-code comments?

**Review by jagpreetsinghsasan — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (repository23) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:63]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (repository23) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:1]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (repository23) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:184]: Some developers might need it during progressing with their integration of this plugin within their project.

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (repository23) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:417]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (repository23) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:707]: insertDetailedTransactionEntry  is always executed , there are transactions like config transactions and not every transaction is related to chaincodeID, but will recheck that

File: packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/fabric-setup-helpers.ts
Comment (repository23) [reply packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/fabric-setup-helpers.ts:1]: it is used in functional tests and we assume that developers might use it during testing and integration process

used in 
cacti\packages\cactus-plugin-persistence-fabric\src\test\typescript\integration\persistence-fabric-functional.test.ts

 88 import {
      enrollAdmin,
      enrollUser,
      getUserCryptoFromWallet,
      } from "./fabric-setup-helpers";

File: tsconfig.base.json
Comment (repository23) [reply tsconfig.base.json:1]: ok restoring back

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:1]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:63]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:417]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:483]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:598]: insertDetailedTransactionEntry is always executed , there are transactions like config transactions and not every transaction is related to chaincodeID, but will recheck that

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:707]: insertDetailedTransactionEntry is always executed , there are transactions like config transactions and not every transaction is related to chaincodeID, but will recheck that

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/web-services/status-endpoint-v1.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/web-services/status-endpoint-v1.ts:54]: ok removing

File: packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/fabric-setup-helpers.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/fabric-setup-helpers.ts:1]: it is used in functional tests and we assume that developers might use it during testing and integration process

used in
cacti\packages\cactus-plugin-persistence-fabric\src\test\typescript\integration\persistence-fabric-functional.test.ts

88 import {
enrollAdmin,
enrollUser,
getUserCryptoFromWallet,
} from "./fabric-setup-helpers";

File: tsconfig.base.json
Comment (barnapa) [reply tsconfig.base.json:1]: ok restoring back

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:138]: It is used during integrations as one of steps for developers to integrate plugin into their project, without setting up all things around like database or ledger.

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:518]: it is styling concept and I used to such view as more readable at this point

File: packages/cactus-cmd-api-server/package.json
Comment (petermetz) [inline packages/cactus-cmd-api-server/package.json:70]: @barnapa Why the downgrade here?

File: .cspell.json
Comment (petermetz) [inline .cspell.json:2]: @barnapa Please eliminate whitespaces from the diff.

File: jest.config.js
Comment (petermetz) [inline jest.config.js:8]: @barnapa This will also enable coverage collection on the CI but for that to be useful it also has to be collected and uploaded from each job and then massaged together (from the different jobs) so that the coverage badge gets updated on codecov.io.
Until all of that happens I suggest to remove this flag for now. You can still use it locally as you see fit.

File: packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts:1108]: @barnapa If we continue down this path the function here will grow to infinity in term of LoC.
I recommend creating a `get-last-block-v1` endpoint instead which then can have a `getLastBlockV1()` method here in the connector that holds nothing but the exact logic what you need.
It also reduces the number of if conditions in the code by 1 which is always a win in my book because it reduces complexity and increases readability.

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/generated/openapi/typescript-axios/.openapi-generator-ignore
Comment (petermetz) [inline packages/cactus-plugin-persistence-fabric/src/main/typescript/generated/openapi/typescript-axios/.openapi-generator-ignore:23]: @barnapa Please use the ignore patterns that all the other packages use as well (you can check any other ignore files of the open api generator for reference)
Once the files are ignored please make sure to remove the from the diff of the PR as well (e.g. git_push.sh, .gitignore, etc.)

File: packages/cactus-plugin-persistence-fabric/package.json
Comment (petermetz) [inline packages/cactus-plugin-persistence-fabric/package.json:3]: @barnapa You'll need to update the version to match the current latest (v1.1.4 as I'm writing this).

The same goes for all the cross-package dependencies declarations here in both `dependecies` and `devDependencies`.

File: packages/cactus-plugin-persistence-fabric/package.json
Comment (petermetz) [inline packages/cactus-plugin-persistence-fabric/package.json:73]: @barnapa Need to update the versions here, please see details above.

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (petermetz) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:138]: @barnapa Any chance that you could call it something like healthcheck instead? I want to keep test code and production code separate and with clear lines between the two. What you describe here is a legitimate use-case for production in my opinion but then I'd avoid anything "test" in its name for that exact reason just so that it doesn't confuse people.

**Review by petermetz — CHANGES_REQUESTED** [review]

**Review by petermetz — CHANGES_REQUESTED** [review]
@barnapa Please fix the merge conflicts via a rebase onto upstream/main

**Comment by petermetz** [comment]
@barnapa Sorry, I misspoke, the latest version isn't 1.1.4 it's 1.2.0

File: packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts:1108]: get-last-block-v1 endpoint in which plugin ? connector or persistence ?

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (jagpreetsinghsasan) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts:518]: But we already have the interface in place, which has the exact same fields. We can simply create an instance for that interface and then use it with just `instance_name.property`. (Lemme know if there are any issues in doing that way, because I would agree to the readable part only if the variables are not just `""`, but in the current case, all of them are empty strings)

File: packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts
Comment (barnapa) [reply packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts:1108]: ok checking on that

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/plugin-persistence-fabric.ts]: @jagpreetsinghsasan Hi, made some changes in last force push by me. Please tell me if this matches what was in your idea :)

File: .cspell.json
Comment (rwat17) [reply .cspell.json:2]: @petermetz Done in my latest force push. But please double check if it is what you wanted. :)

File: jest.config.js
Comment (rwat17) [reply jest.config.js]: @petermetz Deleted, I think it got omitted while refactoring before PR :)

File: packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts
Comment (rwat17) [reply packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts]: @petermetz For now deleted this functionality.

File: packages/cactus-plugin-persistence-fabric/src/main/typescript/generated/openapi/typescript-axios/.openapi-generator-ignore
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/src/main/typescript/generated/openapi/typescript-axios/.openapi-generator-ignore:23]: @petermetz check if it is correct now :)

File: packages/cactus-plugin-persistence-fabric/package.json
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/package.json]: @petermetz Done, please double check :)

File: packages/cactus-plugin-persistence-fabric/README.md
Comment (VRamakrishna) [inline packages/cactus-plugin-persistence-fabric/README.md:39]: "plugins" --> "plugin"

File: packages/cactus-plugin-persistence-fabric/README.md
Comment (VRamakrishna) [inline packages/cactus-plugin-persistence-fabric/README.md:5]: Can you link to the folder containing the GUI tool, just for precision?

File: packages/cactus-plugin-persistence-fabric/package.json
Comment (VRamakrishna) [inline packages/cactus-plugin-persistence-fabric/package.json:23]: I believe this is now "cacti@lists.hyperledger.org". @petermetz Is that right?

File: packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/persistence-fabric-db-client.test.ts
Comment (VRamakrishna) [inline packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/persistence-fabric-db-client.test.ts:126]: Can you insert a blank line above the comment?

File: packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/persistence-fabric-db-client.test.ts
Comment (VRamakrishna) [inline packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/persistence-fabric-db-client.test.ts:220]: Insert blank line above

File: packages/cactus-plugin-persistence-fabric/package.json
Comment (petermetz) [reply packages/cactus-plugin-persistence-fabric/package.json:23]: > I believe this is now "[cacti@lists.hyperledger.org](mailto:cacti@lists.hyperledger.org)". @petermetz Is that right?

@VRamakrishna Yes, that is correct.

File: packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts:1191]: @barnapa No matching strings please. Please declare this as a constant and then also make sure to export it as part of the public API surface so that people who are using it can be sure that they are typing in the correct string because the compiler type-checks it.

File: .cspell.json
Comment (petermetz) [reply .cspell.json:2]: @rwat17 The whitespace is still in the diff, attaching screenshot from my local checkout.


Screenshot from 2023-05-25 17-07-44

**Review by petermetz — CHANGES_REQUESTED** [review]
@barnapa Please 
1. Update the cacti versions to the current latest (which just had a release yesterday)
2. Check out my comment about the magic strings

**Review by VRamakrishna — CHANGES_REQUESTED** [review]
I've requested minor edits around terminology and formatting in my comments.

File: packages/cactus-plugin-persistence-fabric/README.md
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/README.md]: > "plugins" --> "plugin"

done

File: packages/cactus-plugin-persistence-fabric/README.md
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/README.md]: > Can you link to the folder containing the GUI tool, just for precision?

done

File: packages/cactus-plugin-persistence-fabric/package.json
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/package.json]: > Do you want to add "Cacti" to the list?

Done

File: packages/cactus-plugin-persistence-fabric/package.json
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/package.json]: > I believe this is now "[cacti@lists.hyperledger.org](mailto:cacti@lists.hyperledger.org)". @petermetz Is that right?

fixed

File: packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/persistence-fabric-db-client.test.ts
Comment (rwat17) [reply packages/cactus-plugin-persistence-fabric/src/test/typescript/integration/persistence-fabric-db-client.test.ts:127]: > Can you insert a blank line above the comment (better readability)?

done

File: packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts
Comment (rwat17) [reply packages/cactus-plugin-ledger-connector-fabric/src/main/typescript/plugin-ledger-connector-fabric.ts]: @petermetz Changed it as you requested.

**Review by VRamakrishna — APPROVED** [review]

**Review by jagpreetsinghsasan — APPROVED** [review]
LGTM

**Review by petermetz — APPROVED** [review]
@barnapa LGTM

FYI: I've made a couple of changes to the
1. getRuntimeErrorCause function so that it doesn't hide information about exceptions that are not `string` nor `Error` typed
2. dependencies (removed unused ones, added missing declarations)

**Pull request merged**
